### ISSUE TITLE
`dyn Trait<'a>` shouldn't carry `'a` lifetime

### ISSUE THREAD

**Issue by yshui** [rust-lang/rust#104684]
I tried this code:

```rust
fn check<'a, T: Trait<'a> + ?Sized>() {}
trait Trait<'a>: 'static {}

fn check2<'a>() {
    check::<dyn Trait<'a> + 'static>();
}
```

I expected to see this happen: It compiles

`dyn Trait<'a>` should have an lifetime independent from `'a`. i.e. `dyn Trait<'a> + 'b` should have lifetime `'b`. in this particular case, `'b` is `'static`.

Instead, this happened: 

```
error: lifetime may not live long enough
 --> src/lib.rs:5:5
  |
4 | fn check2<'a>() {
  |           -- lifetime `'a` defined here
5 |     check::<dyn Trait<'a> + 'static>();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ requires that `'a` must outlive `'static`
```

### Related

#88904. Since this is another way of creating `dyn Trait: !Trait`, except in this case, I think the unsized coercion should be allowed, and `dyn Trait` does impl `Trait`.

**Comment by Jules-Bertholet** [comment]
[It's an invariant of the type system](https://rust-lang.github.io/rfcs/1214-projections-lifetimes-and-wf.html#outlives-for-projections) that projections from a type can't have a lesser lifetime than that type. In other words, if `T: 'static`, then `T::Some::Arbitary::Path: 'static` must also hold (unless the path involves some non-`'static` lifetime parameter). If `dyn Trait<'a>` was `'static`, this rule could be violated:

```rust
trait Foo {
    type Assoc;
}

impl<'a> Foo for dyn Trait<'a> {
    type Assoc = &'a u32;
}

// `<dyn Trait<'a> as Foo>::Assoc` is not `'static`, so `dyn Trait<'a>` cannot be `'static` either
```

@rustbot label -C-bug C-discussion A-lifetimes A-trait-objects T-types

**Comment by yshui** [comment]
@Jules-Bertholet hmm, ok.

What if we explicitly apply this rule and disallow the impl in this case?

```rust
impl<'a> Foo for dyn Trait<'a> + 'static {
    type Assoc = &'a u32; // error: <dyn Trait<'a> + 'static as Foo>::Assoc must outlive 'static
}
```

I would expect this at least needs an RFC, but does this make sense?

**Comment by Jules-Bertholet** [comment]
@yshui That would be a breaking change, as the impl is currently accepted on stable Rust.

**Comment by yshui** [comment]
true, but that only affects impls explicitly for  `dyn Trait + 'lifetime`, currently this `+ 'lifetime` does nothing, so I doubt there's a lot of code out there written like this.
